### ISSUE TITLE
Maping over array of union type - poor inference

### ISSUE THREAD

**Issue by Andarist** [facebook/flow#4048]
I honestly dont know if Im right, but this one seems easily inferrable and I just cannot make it work. It might be a bug (cannot know that for sure) or maybe there is some reason why this is happening and Im just not aware of the problem in my code.

Would be greatly appreciated if somebody can clarify things up for me.

https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgAeAUgMoDyAcmALxgDeAPmAFYDOcAdgFxjkBGLPAGMMYRgF902fGACSAdQDCpWg2YBLBMLa82GAE4bOAcwA04qZlwEAsnjZsAhiYJ0SFas0UqA3OmEufTAcJwM2PHtHFzcwAAoAWwdnV10wAEEDAycsAB4olLwAPgBKXkzsvI8qItpa+lQASAM8DABXA04wJOjUgDoEpxw4xN4CmJK6hlQwWbBGjSh4hL6tHUmGua25lvbOhlYOHjAyKj7Q8LxE1e02PrYcGA0MOIByAD1XkoBtAEYAXUmUm2c2B2xm22arQ6XQSENmEhKqAkQA

**Comment by rkurbatov** [comment]
https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgAeAUgMoDyAcmALxgDeAPmAFYDOcAdgFxjkBGLPAGMMYRgF9MuAgEkA6gGFStBswCWCYW15sMAJ3WcA5uKnZ8YALJ42bAIbGCdEhWrMFygNzphXPWA49vpseDZ2js5gABQAtrYOTjpgAIL6+vZYADzhiXgAfACUvGkZ2a5U+bRV8RFJAHSx9jjRqGAxsby5kYXVMfRt7UND7Fy8sfWa2mAA-GBkVPVBIXhxk1ps9Ww4MOoY0QDkAHoHhQDaAIwAur3j9aOc4sz0UsPDEr1vqIVAA

this works for me. And agree, union types is the most problematic. It's interesting to check how it behaves in older version. We're trying to switch to the new flow since 0.44 (we're using 0.43.1 currently) but still have a lot of warnings on redux actions with union typing.

**Comment by Andarist** [comment]
@rkurbatov thanks for the answer! I finally managed to make it work, but some things still aint working like expected (see EDIT note in the original post)

**Comment by rkurbatov** [comment]
Yeah, seems like `if (m.iwcs)` and `if (m.iwcs !== undefined)` are equivalent. Looks like a bug.

**Comment by Andarist** [comment]
not rly - `m.iwcs` is not working, while `m.iwcs !== undefined` is the only one making it to work - while other mentioned checks should be OK too

`m.iwcs` is not working, as it shouldnt - because empty string wouldnt go into this block of the code, which is responsible for handling that case

**Comment by toxic-johann** [comment]
bump into same problem here.
[my emample](https://flow.org/try/#0C4TwDgpgBAggxnCBnJB7ATgEWXdBLMYDKAXigG8AfAKCigHMJgAuKAMQFcA7OYPVLgBpaUJE1acefAcLpwBAMzz0O6AIYAjADYRWG1Kh1qhIiFw4BbCOu26o+wxGPVKAX2rVQkKJjXA12Ei4BETopBQ0dABualocdsYgslDyXEoqNjp6BkYmdADu+P622Y7GyWaW1ppZ9jlOXC7unuDQgcGExGS+-u34nWGUsAjIaFg4-aEA3B4K3Lz8XFDoTgAmAlogUAAUqBoAVqwA8gcQvIJQYOioYKxIwPhc9BerEyEYrH3v6ACUn28DCgiAD0wKgeAU2wAhNsAOSFPDFHSw8FLV5BSYYH4-ZZMVRogHTEQQ7bojqhAB0CKR0BIdKg3FeSi4EFWOJWwHxUDJmPQMzoPO+VKKNVpUAUsTE-NxnPQBIx3xmriAA)

flow can't do well in Disjoint unions with exact types
